packages/redaction/src/generic/redact-recursively.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Given an object recurses through it's structure, including inside arrays, and applies
+ * the provided redaction functions. For example:
+ *
+ * ```js
+ * // Replace all (non-id) strings with asterixes
+ * redactRecursively({
+ *  obj: myObject,
+ *  redactString: (str, path) => path.at(-1) === "id" ? str : str.replace(/./g, "*"),
+ * })
+ * ```
+ */
+export const redactRecursively = (
+  rootValue: object,
+  {
+    redactObject,
+    redactArray,
+    redactString,
+    redactNumber,
+    redactBigInt,
+    redactBoolean,
+  }: {
+    /**
+     * If the original object is returned then redactRecursively will recurse inside it,
+     * and continue applying the redaction.
+     *
+     * If a new object is returned (different memory reference) then redactRecursively will
+     * replace the existing object with the new object and not recurse inside it.
+     */
```

Comment:
```suggestion
    /**
     * If the original object is returned then redactRecursively will recurse inside it,
     * and continue applying the redaction.
     *
     * If a new object is returned (different memory reference) then redactRecursively will
     * replace the existing object with the new object and not recurse inside it.
     *
     * In the case where you return the obj passed in unchanged, and so allow redactRecursively
     * to continue recursing inside the object, you should make sure that your redactObject function
     * returns quickly, and does NOT take time proportional to the size of the nested object. For example
     * the below code would be O(n^2) in the size of the object and so extremely unperformant on large objects:
     *
     * ```
     * const redactObject = (obj) => {
     *   const containsSecret = JSON.stringify(obj).includes("secret"); // O(N) where N is size of obj
     *   return containsSecret ? { __redacted__: true } : obj;
     * }
     * const redacted = redactRecursively(aHugeObject, { redactObject }); // O(N^2) (!)
     * ```
     */
```